### PR TITLE
Add Name Domain & Factory

### LINKED ISSUE #1: Name & Address Entity
Code the domain model use Java Records

### LINKED ISSUE #6: Name & Address Factory
Implement factory pattern

### PR DESCRIPTION
**Pull request by leokeke96**
Hi Curstin, i hope you are good today and you are progressing in your code, this is what i have done so far and i wanted to ask in the address entity why City is mentioned that way City : city

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -51,6 +51,11 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.12.0</version>
+        </dependency>
```
Comment (ItsCurstin) [inline pom.xml]: remove this dependency, it is already included in the spring starters

File: src/main/java/za/ac/cput/school_management_grp33/domain/location/Address.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package za.ac.cput.school_management_grp33.domain.location;
+/*
+Address.Java;
+Author: Kevin Lionel Mombo Ndinga(218180500)
+Date: 11 of June 2022;
+ */
+public class Address {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/domain/location/Address.java]: this belongs in the lookup package, not in the location

File: src/main/java/za/ac/cput/school_management_grp33/controller/location/AddressController.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package za.ac.cput.school_management_grp33.controller.location;
+
+public class AddressController {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/controller/location/AddressController.java:4]: this belongs in the lookup package, not in the location

File: src/main/java/za/ac/cput/school_management_grp33/domain/lookup/Name.java [hunk 4]
```diff
@@ -1,4 +1,61 @@
...
+        public Builder copy(Name name) {
+            this.firstName = name.firstName;
+            this.middleName = name.middleName;
+            this.lastName = name.lastName;
+            return this;
+        }
+        public Name build(){
+            return new Name(this);
+        }
+    }
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/domain/lookup/Name.java]: Use Lombok `@Builder`, `@Getter`, `@ToString`, and `@EqualsAndHashCode`. And remove this code.

File: src/main/java/za/ac/cput/school_management_grp33/factory/location/AddressFactory.java [hunk 5]
```diff
@@ -0,0 +1,4 @@
+package za.ac.cput.school_management_grp33.factory.location;
+
+public class AddressFactory {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/factory/location/AddressFactory.java]: this belongs in the lookup package, not in the location

File: src/main/java/za/ac/cput/school_management_grp33/factory/lookup/NameFactory.java [hunk 6]
```diff
@@ -1,4 +1,20 @@
...
+Date: 11 of june 2022;
+ */
 public class NameFactory {
+    public static Name build(String firstName, String middleName, String lastName) {
+        Util.checkStringParam("firstName",firstName);
+        Util.checkStringParam("middleName",middleName);
+        Util.checkStringParam("lastName",lastName);
+
+        return new Name.Builder().firstName(firstName).middleName(middleName).
+                lastName(lastName).build();
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/factory/lookup/NameFactory.java]: You might have to change this once you remove your builder code. Just make sure it compiles.

File: src/main/java/za/ac/cput/school_management_grp33/repository/location/AddressRepository.java [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package za.ac.cput.school_management_grp33.repository.location;
+
+public interface AddressRepository {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/repository/location/AddressRepository.java]: this belongs in the lookup package, not in the location

File: src/main/java/za/ac/cput/school_management_grp33/service/location/AddressService.java [hunk 8]
```diff
@@ -0,0 +1,4 @@
+package za.ac.cput.school_management_grp33.service.location;
+
+public interface AddressService {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/service/location/AddressService.java]: this belongs in the lookup package, not in the location

File: src/main/java/za/ac/cput/school_management_grp33/service/location/impl/AddressServiceImp.java [hunk 9]
```diff
@@ -0,0 +1,4 @@
+package za.ac.cput.school_management_grp33.service.location.impl;
+
+public class AddressServiceImp {
+}
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/service/location/impl/AddressServiceImp.java:4]: this belongs in the lookup package, not in the location. Also, it is `AddressServiceImpl`. You forgot the l at the end.

File: src/main/java/za/ac/cput/school_management_grp33/util/Util.java [hunk 10]
```diff
@@ -1,9 +1,23 @@
...
+import org.apache.commons.lang3.StringUtils;
+
 /**
  * Utility Class, also known as Helper class.
  * It contains a bunch of related methods,
  * so they can be reused across the application.
  */
 public final class Util {
+    public static boolean isEmptyOrNull(String str) {
+        return StringUtils.isEmpty(str);
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/util/Util.java]: once you remove the apache-commons dependency, change this to `return StringUtils.hasText(str);`
It does the same thing

File: src/main/java/za/ac/cput/school_management_grp33/util/Util.java [hunk 11]
```diff
@@ -1,9 +1,23 @@
...
+
+    }
+    public static String setEmptyIfNull(String str) {
+    if(isEmptyOrNull(str)) return StringUtils.EMPTY;
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/util/Util.java]: change the return of this as well to `return "";` remove `StringUtils.EMPTY`

File: src/main/java/za/ac/cput/school_management_grp33/util/Util.java [hunk 12]
```diff
@@ -1,9 +1,23 @@
 package za.ac.cput.school_management_grp33.util;
 
+import org.apache.commons.lang3.StringUtils;
```
Comment (ItsCurstin) [inline src/main/java/za/ac/cput/school_management_grp33/util/Util.java]: replace this line with `import org.springframework.util.StringUtils;`

**Review by ItsCurstin — CHANGES_REQUESTED** [review]
I mentioned a few changes. Don't open a new pull request just push your code to your repository, it will reflect here.

File: pom.xml (hunk 1 above)
Comment (leokeke96) [reply pom.xml]: i removed it but if you can check if what i will do in the util later to see any mistakes

File: src/main/java/za/ac/cput/school_management_grp33/domain/location/Address.java (hunk 2 above)
Comment (leokeke96) [reply src/main/java/za/ac/cput/school_management_grp33/domain/location/Address.java]: ok great i will change it now

File: src/main/java/za/ac/cput/school_management_grp33/factory/lookup/NameFactory.java (hunk 6 above)
Comment (leokeke96) [reply src/main/java/za/ac/cput/school_management_grp33/factory/lookup/NameFactory.java]: how can i change it because i got already errors

**Pull request closed without merging**